lib/lz4hc.c

```diff
@@ -247,34 +248,35 @@ LZ4HC_InsertAndGetWiderMatch (
             /* do nothing */
         } else if (matchIndex >= dictLimit) {
             const BYTE* const matchPtr = base + matchIndex;
+            assert(matchPtr >= lowPrefixPtr);
+            assert(matchPtr < ip);
             assert(longest >= 1);
-            if (LZ4_read16(iLowLimit + longest - 1) == LZ4_read16(matchPtr - delta + longest - 1)) {
+            if (LZ4_read16(iLowLimit + longest - 1) == LZ4_read16(matchPtr - lookBackLength + longest - 1)) {
                 if (LZ4_read32(matchPtr) == pattern) {
                     int mlt = MINMATCH + LZ4_count(ip+MINMATCH, matchPtr+MINMATCH, iHighLimit);
-                    int const back = delta ? LZ4HC_countBack(ip, matchPtr, iLowLimit, lowPrefixPtr) : 0;
+                    int const back = lookBackLength ? LZ4HC_countBack(ip, matchPtr, iLowLimit, lowPrefixPtr) : 0;
                     mlt -= back;
-
                     if (mlt > longest) {
                         longest = mlt;
                         *matchpos = matchPtr+back;
                         *startpos = ip+back;
-                }   }
-            }
+            }   }   }
         } else {   /* matchIndex < dictLimit */
             const BYTE* const matchPtr = dictBase + matchIndex;
             if (LZ4_read32(matchPtr) == pattern) {
+                const BYTE* const dictLowLimit = dictBase + hc4->lowLimit;
```

Comment:
I think this should be given a different name or use `lowLimit` directly. `dictLowLimit` implices it is related to `lowLimit`, which it sort of is. However, `dictLowLimit` might include space that is below `lowLimit`. It is okay in this scenario, since we are only using it for backwards counting, so we already have a fixed offset, but someone might confuse it in the future.

Reply:
I agree, the issue here is using `lowLimit` to mean different things.
Unfortunately, I failed to find a clear enough naming convention to fix that situation...

Reply:
Is it still correct to use `lowLimit` instead of `hc4->lowLimit` here? That way we can keep the dual-meaning consistent.

Reply:
I believe it's not.
`lowLimit` is for initial distance check, from `ip`.
`hc4->lowLimit` is the place we must stop looking backward because there is no more data.
Sometimes, both correspond. But in many cases, they do not.

Note that, the compression impact is extremely limited : one would need an initial match very close to the distance limit (65534 bytes for example) and then multiple bytes of look back. In which case, using` lowLimit` instead of `hc4->lowLimit`, the look-back stage would be limited, and potentially miss a few bytes of compression.

So using `lowLimit` still works, and negatively impact compression ratio by a negligible amount.
But it's not correct. Hence the proposed fix.

Reply:
Changes variables name to `dictStart`

Reply:
Also renamed `lowLimit` to become `lowestMatchIndex`
which seems more accurate.